File: docs/research/spikes/one-gb-runtime-capacity/harness.mjs
import http from "node:http";
import { monitorEventLoopDelay } from "node:perf_hooks";
import fs from "node:fs";

const seconds = Number(process.argv[2] ?? 5);
if (!Number.isFinite(seconds) || seconds < 1) throw new Error("duration must be positive seconds");

const started = Date.now();
const storePath = process.env.WAW_SYNTHETIC_STORE;
if (!storePath || fs.statSync(storePath).size !== 40 * 1024 * 1024) throw new Error("40MiB synthetic store is required");
const store = fs.openSync(storePath, "r+");
const delays = monitorEventLoopDelay({ resolution: 10 });
const seen = new Set();
let applied = 0;
for (let i = 0; i < 10_000; i++) {
  const id = i % 9_000;
  if (!seen.has(id)) {
    seen.add(id);
    applied++;
  }
}

const server = http.createServer((request, response) => {
  if (!["/health", "/status", "/settings"].includes(request.url)) {
    response.writeHead(404).end();
    return;
  }
  response.setHeader("content-type", "application/json");
  response.end(JSON.stringify({ ok: true, applied }));
});

await new Promise((resolve) => server.listen(0, "127.0.0.1", resolve));
delays.enable();
const { port } = server.address();
const latencies = [];
let requests = 0;
let errors = 0;
const endpointRequests = { "/health": 0, "/status": 0, "/settings": 0 };
const endpointErrors = { "/health": 0, "/status": 0, "/settings": 0 };
const rssSamples = [];
const rssTimer = setInterval(() => {
  rssSamples.push(process.memoryUsage().rss / 1024 / 1024);
}, 100);

async function request(path) {
  const before = performance.now();
  try {
    const response = await fetch(`http://127.0.0.1:${port}${path}`);
    const body = await response.json();
    if (!response.ok || body.ok !== true || body.applied !== 9_000) {
      errors++;
      endpointErrors[path]++;
    }
  } catch {
    errors++;
    endpointErrors[path]++;
  }
  requests++;
  endpointRequests[path]++;
  latencies.push(performance.now() - before);
}

const paths = ["/health", "/status", "/settings"];
let tick = 0;
const workloadStarted = performance.now();
let nextBatch = workloadStarted;
while (performance.now() - workloadStarted < seconds * 1000) {
  const buffer = Buffer.alloc(4096, tick % 256);
  const offset = (tick * buffer.length) % (40 * 1024 * 1024 - buffer.length);
  fs.writeSync(store, buffer, 0, buffer.length, offset);
  fs.readSync(store, buffer, 0, buffer.length, offset);
  await Promise.all(Array.from({ length: 5 }, (_, i) => request(paths[(tick + i) % paths.length])));
  tick++;
  nextBatch += 2500;
  await new Promise((resolve) => setTimeout(resolve, Math.max(0, nextBatch - performance.now())));
}

clearInterval(rssTimer);
delays.disable();
await new Promise((resolve) => server.close(resolve));
fs.closeSync(store);
latencies.sort((a, b) => a - b);
rssSamples.sort((a, b) => a - b);
const percentile = (values, fraction) => values[Math.max(0, Math.ceil(values.length * fraction) - 1)] ?? 0;
console.log(JSON.stringify({
  runtime: `node ${process.version}`,
  duration_seconds: (Date.now() - started) / 1000,
  input_events: 10_000,
  duplicate_events: 1_000,
  applied_events: applied,
  requests,
  request_rate_per_second: requests / seconds,
  endpoint_requests: endpointRequests,
  endpoint_errors: endpointErrors,
  errors,
  error_rate: requests ? errors / requests : 1,
  http_p95_ms: percentile(latencies, 0.95),
  scheduler_p99_ms: delays.percentile(99) / 1e6,
  rss_p95_mib: percentile(rssSamples, 0.95),
}));

File: docs/research/spikes/one-gb-runtime-capacity/harness.py
import json
import resource
import sys
import threading
import time
import urllib.request
from concurrent.futures import ThreadPoolExecutor
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

seconds = float(sys.argv[1]) if len(sys.argv) > 1 else 5
if seconds < 1:
    raise ValueError("duration must be positive seconds")

started = time.monotonic()
seen = set()
for event in range(10_000):
    seen.add(event % 9_000)


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path not in ("/health", "/status", "/settings"):
            self.send_error(404)
            return
        body = json.dumps({"ok": True, "applied": len(seen)}).encode()
        self.send_response(200)
        self.send_header("content-type", "application/json")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *_):
        pass


server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
threading.Thread(target=server.serve_forever, daemon=True).start()
latencies = []
errors = 0
requests = 0
paths = ("/health", "/status", "/settings")
scheduler_delays = []
rss_samples = []
monitor_done = threading.Event()


def monitor_scheduler():
    expected = time.monotonic()
    while not monitor_done.wait(0.01):
        expected += 0.01
        now = time.monotonic()
        scheduler_delays.append(max(0, (now - expected) * 1000))
        rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
        rss_samples.append(rss / 1024 / 1024 if sys.platform == "darwin" else rss / 1024)
        expected = now


def request(path):
    before = time.monotonic()
    failed = 0
    try:
        with urllib.request.urlopen(f"http://127.0.0.1:{server.server_port}{path}", timeout=2) as response:
            failed = response.status != 200
            response.read()
    except Exception:
        failed = 1
    return (time.monotonic() - before) * 1000, failed


threading.Thread(target=monitor_scheduler, daemon=True).start()

with ThreadPoolExecutor(max_workers=5) as pool:
    while time.monotonic() - started < seconds:
        batch = list(pool.map(request, (paths[(requests + offset) % 3] for offset in range(5))))
        latencies.extend(item[0] for item in batch)
        errors += sum(item[1] for item in batch)
        requests += 5
        time.sleep(2.5)

monitor_done.set()
server.shutdown()
latencies.sort()
scheduler_delays.sort()
percentile = lambda values, fraction: values[max(0, int(len(values) * fraction + 0.999999) - 1)] if values else 0
rss_samples.sort()
print(json.dumps({
    "runtime": sys.version.split()[0],
    "duration_seconds": time.monotonic() - started,
    "input_events": 10_000,
    "duplicate_events": 1_000,
    "applied_events": len(seen),
    "requests": requests,
    "errors": errors,
    "error_rate": errors / requests if requests else 1,
    "http_p95_ms": percentile(latencies, 0.95),
    "scheduler_p99_ms": percentile(scheduler_delays, 0.99),
    "rss_p95_mib": percentile(rss_samples, 0.95),
}))
